### PR TITLE
support complex input for upsample

### PR DESCRIPTION
**Pull request by mloubout**
The current implementation uses the array input type to convert width/height/.... which creates non-real numbers when the input is complex. This fixes it by using the real type.

### CODE REVIEW

File: src/upsample.jl [hunk 1]
```diff
@@ -180,9 +180,9 @@ function upsample_bilinear_whcn!(output::AbstractArray{T,4}, input::AbstractArra
     out_w, out_h, _, _ = size(output)
     output_slice_size = out_h * out_w
 
-    # T() and // so that we can handle rationals (super slow)
-    width_scale  = T((in_w - 1) // (out_w - 1))
-    height_scale = T((in_h - 1) // (out_h - 1))
+    #real(T)() and // so that we can handle rationals (super slow)
+    width_scale  =real(T)((in_w - 1) // (out_w - 1))
```
Comment (ToucheSir) [inline src/upsample.jl:184]: Do you mind extracting out `real(T)` as `RT` or some like name? Otherwise this LGTM.

Comment (mloubout) [reply src/upsample.jl]: Done

File: test/upsample.jl [hunk 2]
```diff
@@ -172,3 +172,20 @@ end
...
+            @test size(xup)[1:d] == (8,16,24)[1:d]
+            @test real(xup) == interp(real(x), k)
+            @test imag(xup) == interp(imag(x), k)
+
+            xup = interp(x; size=(8,24,48)[1:d])
+            @test size(xup)[1:d] == (8,24,48)[1:d]
+            @test real(xup) == interp(real(x), size=(8,24,48)[1:d])
+            @test imag(xup) == interp(imag(x), size=(8,24,48)[1:d])
+        end
+    end
```
Comment (ToucheSir) [inline test/upsample.jl:190]:
```suggestion
    for (d, method) in zip(1:3, [upsample_linear, upsample_bilinear, upsample_trilinear])
        for (k, interp) in zip((2, ntuple(_ -> 2,  d)), [method, upsample_nearest])
            x = randn(ComplexF32, (4, 8, 12)[1:d]..., 1, 1)
            
            upsize = (8, 16, 24)[1:d]
            xup = interp(x, k)
            @test size(xup)[1:d] == upsize
            @test real(xup) == interp(real(x), k)
            @test imag(xup) == interp(imag(x), k)

            upsize = (8, 24, 48)[1:d]
            xup = interp(x; size=upsize)
            @test size(xup)[1:d] == upsize
            @test real(xup) == interp(real(x), size=upsize)
            @test imag(xup) == interp(imag(x), size=upsize)
        end
    end
```
One more touch-up

Comment (mloubout) [reply test/upsample.jl]: done

**Review by ToucheSir — APPROVED** [review]
Thanks!

**Pull request merged**